assume/markets/clearing_algorithms/contracts.py

```diff
@@ -0,0 +1,659 @@
+        """
+        Checks if a given supply_order fulfills the criteria of the demand_order and vice versa.
+        Used to allow bidding on some policies only to some agents.
+
+        Args:
+            supply_order (Order): the supply order in question
+            demand_order (Order): the demand order in question
+
+        Returns:
+            bool: True if the orders are compatible
+        """
+        s_information = self.registered_agents[supply_order["agent_id"]]
+        d_information = self.registered_agents[demand_order["agent_id"]]
+        return supply_order["eligible_lambda"](d_information) and demand_order[
+            "eligible_lambda"
+        ](s_information)
+
+    def clear(
+        self, orderbook: Orderbook, market_products
+    ) -> (Orderbook, Orderbook, list[dict]):
+        market_getter = itemgetter(
+            "start_time", "end_time", "only_hours", "contract", "evaluation_frequency"
+        )
+        accepted_orders: Orderbook = []
+        rejected_orders: Orderbook = []
+        meta = []
+        orderbook.sort(key=market_getter)
+        for product, product_orders in groupby(orderbook, market_getter):
+            accepted_demand_orders: Orderbook = []
+            accepted_supply_orders: Orderbook = []
+            if product[0:3] not in market_products:
+                rejected_orders.extend(product_orders)
+                # log.debug(f'found unwanted bids for {product} should be {market_products}')
+                continue
+
+            accepted_product_orders = []
+
+            product_orders = list(product_orders)
+            demand_orders = list(filter(lambda x: x["volume"] < 0, product_orders))
+            supply_orders = list(filter(lambda x: x["volume"] > 0, product_orders))
+
+            # generation
+            supply_orders.sort(key=lambda i: i["price"])
+            # demand
+            demand_orders.sort(key=lambda i: i["price"], reverse=True)
+            dem_vol, gen_vol = 0, 0
+            # the following algorithm is inspired by one bar for generation and one for demand
+            # add generation for currents demand price, until it matches demand
+            # generation above it has to be sold for the lower price (or not at all)
+            for demand_order in demand_orders:
+                if not supply_orders:
+                    # if no more generation - reject left over demand
+                    rejected_orders.append(demand_order)
+                    continue
+
+                dem_vol += -demand_order["volume"]
+                to_commit: Orderbook = []
+
+                while supply_orders and gen_vol < dem_vol:
+                    supply_order = supply_orders.pop(0)
+                    if supply_order["price"] <= demand_order[
+                        "price"
+                    ] and self.check_working(supply_order, demand_order):
+                        # TODO this might match a different demand, if different policies are available
```

Comment:
should this TODO be adressed directly now or is this a work for the future?

Reply:
This is actually not needed at all as the orders are grouped by the `market_getter`, which includes grouping by contract. So bids with different contracts can never be matched together.
So this is not an issue at all.
Thanks!